### PR TITLE
Resolving Env vars for build to work on local dataset and elasticsearch dataset

### PR DESCRIPTION
**Pull request by zmhassan**
When building the image we could use s2i binary instead of requiring us to use openshift to build the docker image. Also provided environment variables so users can use env vars instead of passing`--mode train --modelstore s3`

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -36,14 +37,14 @@ def _main():
     args = parser.parse_args()
     # Allow users to pick storage location we may want to
     # store our models in google cloud storage or azure instead of s3 in the future
-    if args.modelstore=="s3":
+    if args.modelstore=="s3" or os.getenv("LAD_MODEL_STORE")== "s3":
```
Comment (durandom) [inline app.py]: missing space ` ` after `MODEL_STORE")`

File: build-s2i.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
Comment (durandom) [inline build-s2i.sh]: as said, a good alternative would be a Makefile. See https://github.com/AICoE/prometheus-flatliner/blob/master/Makefile for a cool one that also reads in `.env` - this way you don't have to specify the 👇 variables in that shell script here.

File: .env.example [hunk 3]
```diff
@@ -20,3 +20,9 @@ LADI_THRESHOLD=1.5 # float tunes the anomaly threshold based on the max value of
 
 LOCAL_DATA_FILE=validation_data/Hadoop_2k.json
 LOCAL_RESULTS_FILE=/tmp/local-results.txt
+LAD_S3_KEY=<REPLACE_WITH_S3_KEY> # The s3 key
+LAD_S3_SECRET=<REPLACE_WITH_S3_SECRET> # The s3 secret
+LAD_S3_HOST=http://localhost:8080 # The s3 endpoint url
+LAD_S3_BUCKET=<REPLACE_WITH_BUCKET_NAME> # The bucket name for s3
+LAD_MODEL_STORE=s3 # If s3 is not provided then it will write model files just to disk and not sync with s3
```
Comment (durandom) [inline .env.example]: tbh, I would prefer an `.env.example` that just works if copied to `.env`. How about making the default empty and the above too so that the upload to S3 doesnt happen?

File: app.py (hunk 1 above)
Comment (zmhassan) [reply app.py]: I'm removing this in another PR. I'm going to have this setup in config.py

File: .env.example (hunk 3 above)
Comment (zmhassan) [reply .env.example]: Good point. I'll leave it blank.

File: app.py (hunk 1 above)
Comment (zmhassan) [reply app.py]: Okay I've removed this and put it in config.py

File: build-s2i.sh (hunk 2 above)
Comment (zmhassan) [reply build-s2i.sh]: Looks good. I can incorporate that here.

File: anomaly_detector/config.py [hunk 4]
```diff
@@ -51,9 +51,9 @@ class Configuration():
     MODEL_PATH = ""
     W2V_MODEL_PATH_CALLABLE = join_w2v_model_path
     W2V_MODEL_PATH = ""
-    MODEL_STORE= ""
+    MODEL_STORE= os.getenv("LAD_MODEL_STORE","")
```
Comment (vpavlin) [inline anomaly_detector/config.py]: You don't need to do this - just define the name of the config option (`MODEL_STORE`) and the `Configuration.load_from_env` will automatically match the name to env variable and set it.

Comment (zmhassan) [reply anomaly_detector/config.py]: @vpavlin what about `LOCAL_DATA_FILE` and `LOCAL_RESULTS_FILE` ? Those seem to have been set with env vars.

Comment (durandom) [reply anomaly_detector/config.py]: that's right. You should make all setting consistent. 
Looking at the code below, you'll want all environment variables prefixed with `LAD` in `.env.example`. 
For testing in travis or locally you might want to use another prefix as `LAD_TEST` - but not sure. Seems [here](https://github.com/AICoE/log-anomaly-detector/blob/master/tests/test_validation_data.py#L5) you're using `LAD` but in `.env.example` you've been using `LADT`

Comment (zmhassan) [reply anomaly_detector/config.py]: The original file had that but I think its a typo. I'll remove that extra 'T'

Comment (zmhassan) [reply anomaly_detector/config.py]: Some of the env vars that are in this file are different from those provided in:
https://github.com/AICoE/log-anomaly-detector/blob/master/openshift/log-anomaly-detector.app.yaml#L87-L126

```
- name: LAD_STORAGE_BACKEND
            - name: LAD_MODEL_DIR
            - name: LAD_MODEL_FILE
            - name: LAD_W2V_MODEL_FILE
            - name: LAD_TRAIN_TIME_SPAN
            - name: LAD_TRAIN_MAX_ENTRIES
            - name: LAD_TRAIN_ITERATIONS
            - name: LAD_TRAIN_UPDATE_MODEL
            - name: LAD_PARALLELISM
            - name: LAD_INFER_TIME_SPAN
            - name: LAD_INFER_MAX_ENTRIES
            - name: LAD_INFER_LOOPS
            - name: LAD_INFER_ANOMALY_THRESHOLD
            - name: LAD_ES_ENDPOINT
            - name: LAD_ES_INPUT_INDEX
            - name: LAD_ES_TARGET_INDEX
            - name: LAD_ES_QUERY
            - name: LAD_ES_CERT_DIR
```

**Comment by zmhassan** [comment]
In this PR I also include code coverage report: 
https://travis-ci.org/AICoE/log-anomaly-detector/builds/491745210
#62

**Comment by durandom** [comment]
> In this PR I also include code coverage report:
> https://travis-ci.org/AICoE/log-anomaly-detector/builds/491745210
> #62

For future PRs, please do always one thing per PR. Makes it easier to review

**Comment by zmhassan** [comment]
Sounds good.

**Comment by zmhassan** [comment]
Hi there,

Just checking back its been 2 weeks. @vpavlin Did you get a chance to review this?
Can this get merged?

**Comment by zmhassan** [comment]
@vpavlin I can quickly fix that and follow the convention. I know it will be better .

**Comment by zmhassan** [comment]
@vpavlin Kindly review this PR. I think it's ready. I've applied the changes you requested.

File: anomaly_detector/config.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 import os
 import distutils
-
+from distutils import util
```
Comment (vpavlin) [inline anomaly_detector/config.py]: Why did you add this import? Does not seem to be used (the import on line 5 is used)

File: anomaly_detector/config.py [hunk 6]
```diff
@@ -51,9 +48,9 @@ class Configuration():
     MODEL_PATH = ""
     W2V_MODEL_PATH_CALLABLE = join_w2v_model_path
     W2V_MODEL_PATH = ""
-    MODEL_STORE= ""
+    MODEL_STORE=  ""
```
Comment (vpavlin) [inline anomaly_detector/config.py]: There is added space, one is enough:)

File: anomaly_detector/storage/local_storage.py [hunk 7]
```diff
@@ -81,10 +81,13 @@ class LSConfiguration(Configuration):
...
-    LS_OUTPUT_PATH = Configuration.LOCAL_RESULTS_FILE
+    LS_OUTPUT_PATH = ""
 
-    def __init__(self):
+    def __init__(self,config):
         """Initialize configuration."""
-        self.load()
+        self.load_from_env()
+        self.LS_INPUT_PATH=config.DATA_FILE
+        self.LS_OUTPUT_PATH = config.RESULTS_FILE
```
Comment (vpavlin) [inline anomaly_detector/storage/local_storage.py]: What is this about? Why do you do this here?

Comment (durandom) [reply anomaly_detector/storage/local_storage.py]: seems to be a mismatch of config variable and environment variable naming. Either use `DATA_FILE` or `LS_INPUT_PATH`.

**Comment by durandom** [comment]
@zmhassan 

* please add new changes as a new commit, this makes it easier to review the changes since the last review. Now I have to go through the complete change again.
* please only change things that concern *one* topic. This is even more important, as it's hard for a reviewer to understand the reasoning of the changes. The topic of this PR is "Adding script to build s2i and allowing for more env vars to setup mo…", whats "mo..." ? Why is there something about codecoverage? 

[Here](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/) is a great post about PRs, inspired by [this guide](https://github.com/thoughtbot/guides/tree/master/code-review) which we used in the manageiq community.

**Comment by zmhassan** [comment]
No problem. I can follow this advice. I understand.

File: anomaly_detector/storage/local_storage.py (hunk 7 above)
Comment (zmhassan) [reply anomaly_detector/storage/local_storage.py]: I've reverted to our original LOCAL_DATA_FILE and LOCAL_RESULTS_FILE

File: anomaly_detector/config.py (hunk 4 above)
Comment (zmhassan) [reply anomaly_detector/config.py]: I think we should get rid of .env file and just add proper documentation on what the fields mean. I could add a very simple .env file with basic fields to train locally

Comment (zmhassan) [reply anomaly_detector/config.py]: Lets move this to another PR. I'll open an issue for this now https://github.com/AICoE/log-anomaly-detector/issues/69

File: build-s2i.sh (hunk 2 above)
Comment (zmhassan) [reply build-s2i.sh]: This file has been removed.

**Comment by durandom** [comment]
@zmhassan please let us know if this is ready to review again

**Comment by zmhassan** [comment]
Yes. Its ready for review. I've address all issues.

File: anomaly_detector/storage/local_storage.py [hunk 8]
```diff
@@ -81,10 +81,13 @@ class LSConfiguration(Configuration):
...
     # Path to a file where the results will be stored. Results will be printed to STDOUT if empty
-    LS_OUTPUT_PATH = Configuration.LOCAL_RESULTS_FILE
+    LS_OUTPUT_PATH = ""
 
-    def __init__(self):
+    def __init__(self,config):
         """Initialize configuration."""
-        self.load()
+        self.load_from_env()
+        self.LS_INPUT_PATH=config.LOCAL_DATA_FILE
```
Comment (durandom) [inline anomaly_detector/storage/local_storage.py]: If you rename this config and the other to `LS_INPUT_PATH` you don't have to make this assignment. Just keep the ENV variables and private variable names in sync

File: anomaly_detector/storage/local_storage.py [hunk 9]
```diff
@@ -81,10 +81,13 @@ class LSConfiguration(Configuration):
     """Configuration for local storage."""
 
     # Path to a file containing input data. If the value is '-' the input is expected on STDIN
-    LS_INPUT_PATH = Configuration.LOCAL_DATA_FILE #"validation_data/verification_data.json"
+    LS_INPUT_PATH = ""
```
Comment (durandom) [inline anomaly_detector/storage/local_storage.py]: I guess these lines can be removed completely then...

File: anomaly_detector/config.py [hunk 10]
```diff
@@ -78,6 +73,16 @@ class Configuration():
     # Maximum number of entries to be loaded for inference
     INFER_MAX_ENTRIES = 78862
 
+    # S3 credentials for storing model up to s3 post training.
+    S3_KEY = ""
+    S3_SECRET = ""
+    S3_HOST = ""
+    S3_BUCKET = ""
+    # local test dataset
+    LOCAL_DATA_FILE = ""
```
Comment (durandom) [inline anomaly_detector/config.py]: so this should be name `LS_INPUT_PATH` right?

File: anomaly_detector/storage/local_storage.py (hunk 8 above)
Comment (zmhassan) [reply anomaly_detector/storage/local_storage.py]: @durandom Good point. Yes. Going to try this

Comment (zmhassan) [reply anomaly_detector/storage/local_storage.py]: I've removed this code. We inject configs in a better way.

File: anomaly_detector/storage/local_storage.py (hunk 9 above)
Comment (zmhassan) [reply anomaly_detector/storage/local_storage.py]: This whole class is gone. No need for it. LSConfiguration

File: anomaly_detector/config.py (hunk 10 above)
Comment (zmhassan) [reply anomaly_detector/config.py]: @durandom yup I renamed that to the name you suggested.

**Comment by zmhassan** [comment]
@durandom Yup I addressed all issues for this PR. I'm just waiting for the build to complete then we can do a final review and maybe merge.

File: anomaly_detector/storage/local_storage.py (hunk 8 above)
Comment (zmhassan) [reply anomaly_detector/storage/local_storage.py]: Done

**Comment by durandom** [comment]
@zmhassan thanks, it's much better than.
